src/shared/bootspec.c

```diff
@@ -1441,9 +1443,13 @@ int show_boot_entries(const BootConfig *config, JsonFormatFlags json_format) {
                         if (r < 0)
                                 return log_oom();
 
-                        json_variant_dump(v, json_format, stdout, NULL);
+                        r = json_variant_append_array(&array, v);
+                        if (r < 0)
+                                return log_oom();
                 }
 
+                json_variant_dump(array, json_format, NULL, NULL);
```

Comment:
we had discussions about this before. passing json objects as infinite stream is actually well supported, tools such as "jq" for example process this just fine.

generally, if stream stuff ("infinite stuff") should *not* be wrapped in a json array, as this forces readers into buffering them in full before parsing them and then end up with a humungous object in memory.

it's one of the reasons we have json_parse_continue() btw, so that we can parse a string buffer for one object, and then continue, reading the next.

Reply:
Well, OK. But this doesn't really matter for the cases in this PR. There is no reason not to generate a valid JSON string that can be easily processed in any tool without fuss.

Reply:
for the userdbctl the user database might have a bazillion entries. and bootctl might too, who knows

Reply:
and as mentioned, most tools are just fine reading multiple whitespace (which includes newline) separated json objects. jq for example. it's quite commonly done.

Reply:
I tested `userdbctl` with my patch and 30k records, and it's unbearably slow with my patch. I guess that our code is O(n²) when appending to an array. I think this is terrible and we need to fix this, regardless of the changes here.

Let's assume that the code is fixed to be reasonably fast with a normal array. The question is what to write by default. I think it's fine if we have a streaming mode where we write output like a sequence of objects as we do now. But I think that we should write a conforming JSON array of objects by default. This is because it's just more useful to users. Generally, the purpose of JSON output is to pipe it into some other program or save to a file and load as a JSON string. Right now the user has to accumulate the inputs and build an array themselves, and this just doesn't seem to be the right default.

Reply:
I think your assumption that "a series of whitespace separate JSON objects" was something users weren't accustomed to is simply not right.

Again, there's a reason why tools such as "jq" actually support streams with JSON objects just fine, because it's common.

Reply:
and also: we might be slow with writing out 30k as an array, but so will be any code that tries to parse that. hence we shouldn't do it.

json arrays are good for "short", bounded stuff.

for "large", conceptually unbounded stuff use a stream of newline separated json objects

Reply:
> and also: we might be slow with writing out 30k as an array, but so will be any code that tries to parse that. hence we shouldn't do it.

I actually tested this. The parsing is very fast, and this is not an issue.

> json arrays are good for "short", bounded stuff.
> 
> for "large", conceptually unbounded stuff use a stream of newline separated json objects

Yes, I don't disagree. But bootctl / userdbctl output is not unbounded. It's a few dozen lines when querying specific users, and a few thousands records when querying a big system.

Reply:
I think python supports the format just fine:

```
userdbctl --json=short > x.json
python3 <<EOF
import json
for l in open("x.json"):
        print(json.loads(l))
EOF
```

seems to work perfectly here, is robust, correct, efficient, and super easy on the eye. And works the same for bootctl's json output and journalctl's and so on.

Reply:
This perfectly shows what I was trying to say. If the goal is to just print the text to the terminal, the existing approach if just fine. But that's not the reason why we provide json output mode. As soon as you try to actually use this, things become annoying:

```ipython
>>> !userdbctl --json=short >x.json
# The obvious approach
>>> import json
>>> json.load(open('x.json'))
JSONDecodeError: Extra data: line 2 column 1 (char 316)
# Oh, so we need to do special mangling of the input data to read it.
# Maybe a bit better than csv, but not by much
>>> arr = []
>>> for l in open('x.json'):
...     arr.append(json.loads(l))
>>> print(arr)
```
Obviously, the input can still be read. But it's not JSON. The whole freakin' reason for a standarized interchange format is that it's trivially readable without manual processing.

Reply:
I am fine with adding an additional mode that blows this up into a json array, but this should be optional, and be "decoration", i.e. not actually create an object in memory.

btw, because this is such a common thing there's apparently also https://pypi.org/project/jsonstream/

And python's native json toolset support this too btw, try:

```
userdbctl --json=short | python3 -m json.tool --json-lines
```

From that I actually learnt that the most pythonesque way to read such a series of json objects is via:

```python
import python
print(list(json.loads(line) for line in open("x.json")))
```

which doesn't look that bad to me